Update to latest xapian

ojwb opened the pull request:
Still working on this, but opening a PR so we can see what CI thinks about it.

The tests pass locally with commit 6e12ad7cd2c4b89fb5e982700553bd54ec0d1261 from Xapian git master.

ojwb committed:
Update instructions for running meson

Fixes meson warning:

WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.

ojwb committed:
Update meson version requirement in metadata

Fixes meson warning:

xapian-glib/meson.build:6: WARNING: Project targets '>= 0.44.0' but uses feature introduced in '0.50.0': install arg in configure_file.

ojwb committed:
Drop volatile from argument to g_once_init_enter()

This fixes a compile error:

../xapian-glib/xapian-enums.cc: In function ‘GType xapian_database_action_get_type()’:
/usr/include/glib-2.0/glib/gatomic.h:131:19: error: argument 2 of ‘__atomic_load’ must not be a pointer to a ‘volatile’ type

The example code in the documentation for g_once_init_enter() doesn't
use `volatile`:

https://docs.gtk.org/glib/type_func.Once.init_enter.html

ojwb committed:
Compile as C++17

This is now the minimum C++ standard required by xapian-core.

ojwb committed:
Fix missing braces on if-s

Fixes:

../xapian-glib/xapian-database.cc:93:3: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]

ojwb committed:
Avoid passing NULL to Xapian::sortable_unserialise()

This has never been valid, but now results in a compile-time error
rather than undefined behaviour at runtime.  Fortunately this code
is only exercised if an invalid XapianDocument is passed in.

ojwb committed:
Fix reversed conditionals in pre-conditions

ojwb committed:
Remove duplicate C++ flag

ojwb committed:
Stop using GCC C-only warnings

These options are not supported by GCC when compiling C++:

-Werror=implicit : implicit return types are always an error in C++.

-WWerror=pointer-to-int-cast : casting a pointer to a narrower integer
type is always an error in C++ (casting to a wider integer isn't, but
is a less problematic case at least).

Meson isn't smart enough to determine that these are only actually
supported for C (it does a test run of the C++ compile which generates
a warning rather than failing so it thinks they're accepted) so they get
used for the build and every C++ file compiled triggers two warnings:

Fixes:

cc1plus: warning: ‘-Werror=’ argument ‘-Werror=implicit’ is not valid for C++
cc1plus: warning: ‘-Werror=’ argument ‘-Werror=pointer-to-int-cast’ is not valid

ojwb committed:
Wrap Xapian::DatabaseNotFoundError

Fixes testsuite failure.

ojwb committed:
Clean up whitespace to match .editorconfig

Fix trailing whitespace on lines and use of tabs for indentation
(which all seemed to be expecting tabstop of 8, but .editorconfig
sets 2).

ojwb committed:
Document Xapian ≥ 1.5.0 is required

This has actually been the requirement since
ee4817331d7d8b23fddbaccf0fd0f7dfd32f0704 in 2018.

ojwb committed:
Fix PACKAGE_URL to point to a working page

The old URL gives 404.

ojwb committed:
Consistently use https for URLs which support it

ojwb committed:
Avoid unnecessary explicit std::string construction

These were never actually needed as std::string allows implicit
construction from const char*, but now the Xapian C++ API takes
std::string_view instead of std::string in many places these are
actually counter-productive and removing them allows the string
to be passed into Xapian without being copied.

ojwb committed:
Enable -Wuninitialized and -Wshadow

These were explicitly turned off, but the code doesn't trigger them
and it seems better to keep it that way.

History shows they were added with the switch to meson, so probably
just copied from another project where suppressing these warnings
was necessary.

ojwb committed:
Fix wrong return type in doc comment

ojwb committed:
Fix out of date documentation comment

Document::remove_posting() will now remove the term if it removes the
final position and the wdf becomes zero.

eos-shell-build commented:
I don't recognize the submitter as a member of @endlessm. Can somebody in @endlessm verify the patch? See https://phabricator.endlessm.com/w/software/build/ci-integration/#pull-requests for more details.

eos-shell-build commented:
I don't recognize the submitter as a member of @endlessm. Can somebody in @endlessm verify the patch? See https://phabricator.endlessm.com/w/software/build/ci-integration/#pull-requests for more details.

ojwb committed:
Wrap missing newer Xapian::Error subclasses

ojwb committed:
Remove lingering code to support Xapian < 1.4

ojwb committed:
Wrap missing newer QueryParser flags

ojwb committed:
Wrap missing newer Query ops

ojwb committed:
Map unknown query ops to Xapian::Query::OP_INVALID

Previously we were using OP_AND, but we now have OP_INVALID which
seems a better choice - if xapian-glib is built without assertions
OP_AND might quietly appear to work whereas OP_INVALID will trigger
a Xapian exception if run as a query.

ojwb commented:
@ramcq Please can you add me to the allowlist here too?

Also, following the CI link from https://github.com/endlessm/libdmodel/pull/21#event-21150270029 I get a login dialog on ci.endlessos.org but it doesn't like my "olly" login details from ci.endlessm-sf.com:8080 (which I'm not surprised by as I last used those 9 years ago).  Perhaps this will just pass in CI though (tests pass locally), but I might need an account there.  Or someone could email me the logs if it's a pain to setup.

ramcq commented:
add to allowlist

ramcq commented:
@ojwb If the libdmodel situation is anything to go by, the reason the tests failed there was the infrastructure (both upstream and on our side) around building the SDK with the tooling from 2019 has somewhat bitrotted. We have a contractor @theevilskeleton working on bringing it back to life with a slightly simpler flatpak-builder approach. If it's just the SDK build as CI, rather than something specific to this module, we can ignore for now!

ramcq commented:
(looks like the same kind of job, trying to build the whole SDK, expect it to fail soon)

ojwb committed:
xapian_enquire_set_query: Document 0 for qlen

This uses the query's natural length which is what you usually want.

ojwb commented:
> If it's just the SDK build as CI, rather than something specific to this module, we can ignore for now!

OK, thanks.  The xapian-glib testsuite passes locally after these changes so failures in CI are almost certainly bitrot there (it'll at least need updating to use a newer commit from xapian's git repo).

ramcq commented:
Do you suggest a specific commit, or are we good with the latest stable release?

ojwb commented:
> Do you suggest a specific commit, or are we good with the latest stable release?

As above, I'm using commit 6e12ad7cd2c4b89fb5e982700553bd54ec0d1261 so that's what I'd recommend for now at least.

We decided that xapian-glib shouldn't aim to support Xapian 1.4.x, which is the current stable release series (2.0.x should begin early next year) - that commit is the current HEAD of Xapian git master (which is in good shape as we're zeroing in on being able to release it as stable).

ojwb committed:
Fix xapian_mset_iterator_prev()

The first iteration moved the iteration to Xapian's "end" leaving it
in a position that can't be validly dereferenced.  If that bug didn't
stop it being used at all then the end condition was also wrong and
it would indicate iteration was over one step early.

ojwb committed:
Fix tests to work if test db already exists

If a testcase fails it may leave behind its test database, causing
a subsequent run to fail because it can't create a new database.

Specifying XAPIAN_DATABASE_ACTION_CREATE_OR_OVERWRITE instead of
XAPIAN_DATABASE_ACTION_CREATE solves this because Xapian will
overwrite an existing database instead of failing.

ojwb committed:
Use Xapian::iterator_rewound()

This avoids constructing a temporary Xapian::MSetIterator object.

ojwb committed:
xapian_enquire_set_query(): Always call to Xapian

Don't optimise away the call to Xapian if the query object is the one
already set as qlen may be different.  Xapian::Enquire::set_query() is
very cheap as it just takes and stores a references to the query object
and stores an integer.

(No regression test because I can't see how to write one using only
the parts of Xapian's C++ API which xapian-glib currently wraps, so
this may well currently be a latent bug - I spotted it from looking
at the code.)

ojwb committed:
Clarify docs of XAPIAN_DATABASE_ACTION_*

ojwb committed:
Wrap XAPIAN_STEM_STRATEGY_STEM_SOME_FULL_POS

ojwb committed:
Wrap missing new TermGenerator flags

ojwb commented:
I think this is ready for review.  The changes fall into several categories:

* Make xapian-glib build with current Xapian git master (selecting C++17 and fixing some bugs that stopped it building)
* Wrap missing new enums/constants where we already wrap that type of constant
* Fix some bugs and rough edges I noticed along the way
* Eliminate unnecessary explicit `std::string` creation to avoid string copies in many cases

I have added regression and feature tests where I could see a way to - in some cases it doesn't seem to be possible without wrapping more methods/classes from the Xapian C++ API.

I'm not sure if it's actually a good idea to be selecting C++17 explicitly - the pattern in modern C++ compiler development now seem to be to default to a new C++ standard version once support for it is solid, and the explicit selection of C++11 was part of the reason it wasn't building with current Xapian git master.  If the GCC version in the environment you're deploying in defaults to at least C++17 you could just drop that.

I don't see a way in meson to select "C++17 unless the compiler defaults to newer" - it seems you can only specify a list of standards to try until one works so you'd need to put `cpp_std=c++23,c++20,c++17` and then keep adding new C++ standards as they come out, and that doesn't take into account what the compiler defaults to, only what it supports selecting via `-std=c++23` or similar.

I guess the meson design logic here is that the code is written for a particular C++ standard so it's better to select that.

TheEvilSkeleton commented on README.md:
```diff
@@ -12,7 +12,7 @@ the GObject Introspection API.
 Xapian-GLib, unsurprisingly, requires:
 
- * Xapian ≥ 1.3.4
+ * Xapian ≥ 1.5.0
```
As the commit message mentions, https://github.com/endlessm/xapian-glib/commit/ee4817331d7d8b23fddbaccf0fd0f7dfd32f0704 depends on 1.5.0, but the [official website](https://xapian.org/) states that the latest version is 1.4.30. Am I missing or is that unintentional?

TheEvilSkeleton replied:
Odd, the milestone was removed - https://trac.xapian.org/milestone/1.5.0 ([archive](https://web.archive.org/web/20241206043030/https://trac.xapian.org/milestone/1.5.0))

ramcq replied:
Addressed above, 1.5.0 isn't out yet but we're targetting HEAD ahead of the 1.5.0 release:

> We decided that xapian-glib shouldn't aim to support Xapian 1.4.x, which is the current stable release series (2.0.x should begin early next year) - that commit is the current HEAD of Xapian git master (which is in good shape as we're zeroing in on being able to release it as stable).

TheEvilSkeleton replied:
Ah, whoops, really sorry for the noise :sweat_smile:

ojwb replied:
By Xapian's current versioning, 1.5.0 is a development release.  For past release series we've actually made such releases (e.g. 1.3.4 in the lead up to 1.4.0) but these days they seem less useful as it seems more expected to just build from git if you want to test the latest code, so there's unlikely to actually be a 1.5.0 release (or if there is it'll be a 2.0.0 release candidate), but 1.5.0 is what current git master reports as its version so kind of makes sense here.

The 1.5.0 milestone was created back when we assumed we'd be making development releases, but really the tickets on that milestone need to be dealt with before 2.0.0 so I renamed it recently.

ojwb replied:
I've pushed a tweak to note the commit hash we're testing with in the README.

ojwb committed:
Note xapian git hash we're testing with

ojwb committed:
Remove XAPIAN_CHECK_VERSION_INTERNAL macro

This duplicates the functionality of the XAPIAN_AT_LEAST macro
provided by Xapian since 1.4.2 (except that XAPIAN_AT_LEAST is
more robust and does not assume that the minor version and
revision are both always < 100).

This macro isn't currently used, so leave a comment in place to
point people to XAPIAN_AT_LEAST.

ramcq commented on xapian-glib/xapian-enums.cc:
```diff
@@ -68,10 +68,12 @@ XAPIAN_GLIB_DEFINE_ENUM_TYPE (XapianError, xapian_error,
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_INVALID_OPERATION, "invalid-operation")
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_UNIMPLEMENTED, "unimplemented")
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE, "database")
+  XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE_CLOSED, "database-closed")
```
I think if we insert new enum values in the middle, we're changing the ABI - ie the constants compiled in to any programs who are testing the error values become invalid. We could move these new enum values to the end, which is less likely to cause such chaos, or we could bump the SOVER. I think for Endless' use case it doesn't much matter; we are only ever shipping this library bundled together with its dependents, but for anyone else using it, it would be polite to do one or the other.

ramcq commented on xapian-glib/xapian-enums.cc:
```diff
@@ -68,10 +68,12 @@ XAPIAN_GLIB_DEFINE_ENUM_TYPE (XapianError, xapian_error,
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE_CREATE, "database-create")
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE_LOCK, "database-lock")
   XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE_MODIFIED, "database-modified")
+  XAPIAN_GLIB_DEFINE_ENUM_VALUE (XAPIAN_ERROR_DATABASE_NOT_FOUND, "database-not-found")
```
Same enum insertion in the middle.

ramcq commented on xapian-glib/xapian-private.h:
```diff
@@ -3,14 +3,9 @@
-#define XAPIAN_CHECK_VERSION_INTERNAL(maj,min,rev) \
-   (_XAPIAN_CUR_VERSION >= _XAPIAN_VERSION_ENCODE_INTERNAL(maj,min,rev))
+/* The old XAPIAN_CHECK_VERSION_INTERNAL(maj,min,rev) macro which was defined
+ * here has been removed.  Use Xapian's own XAPIAN_AT_LEAST(maj,min,rev)
```
Nit: typo in this commit message!

ramcq reviewed (changes requested):
One commit message typo (XAPIAN_AT_LEAT) and I think we need to insert the new errors into the enum at the end. Otherwise, looks great to me; trying to get another actually active engineer in Endless to take a look. Thanks! :)

ojwb committed:
Fully alphabetise XAPIAN_ERROR_* in doc comment

ojwb commented:
All makes sense and addressed.

In the docs it makes more sense to fully alphabetise the `XAPIAN_ERROR_*` codes so I've done that too.

I also noticed missing defines for two of the new `XAPIAN_ERROR_*` codes so I've added them in:

```
--- a/xapian-glib/xapian-enums.h
+++ b/xapian-glib/xapian-enums.h
@@ -118,7 +118,9 @@ typedef enum {
 
 /* Allow #ifdef checks for more recently added enum values. */
 #ifndef __GTK_DOC_IGNORE__
+#define XAPIAN_ERROR_WILDCARD XAPIAN_ERROR_WILDCARD
 #define XAPIAN_ERROR_DATABASE_NOT_FOUND XAPIAN_ERROR_DATABASE_NOT_FOUND
+#define XAPIAN_ERROR_DATABASE_CLOSED XAPIAN_ERROR_DATABASE_CLOSED
 #endif
 
 XAPIAN_GLIB_AVAILABLE_IN_2_0
```

I had to rewrite history to fix the commit message typo, so I also folded the other changes into the relevant commits.  You can see the diff between what you reviewed and the force-pushed branch at: https://github.com/endlessm/xapian-glib/compare/69af7018fd3dcb8345444f99c23faf36fb6add90..ab2993444d1dc09933741cc239713eb0398f53bc

wjt commented on xapian-glib/tests/database.c:
```diff
@@ -66,7 +66,7 @@ database_writable_new (void)
   GError *error = NULL;
   XapianWritableDatabase *db =
     xapian_writable_database_new ("doesexist",
-                                  XAPIAN_DATABASE_ACTION_CREATE,
+                                  XAPIAN_DATABASE_ACTION_CREATE_OR_OVERWRITE,
```
I think this is the wrong solution to the problem. The test cases should use unique temporary directories. GLib has an easy way to do this: pass `G_TEST_OPTION_ISOLATE_DIRS` https://docs.gtk.org/glib/const.TEST_OPTION_ISOLATE_DIRS.html to `g_test_init`. Then, inside the test, you can use functions like `g_get_home_dir` https://docs.gtk.org/glib/func.get_home_dir.html and get back a path to a temporary subdirectory. The subdir may not exist so you have to create it. I normally use `g_test_add` (rather than `g_test_add_func`) and pass it a `fsetup` function that creates the directory in question.

ojwb replied:
That would probably be better, but sounds rather fiddly to get right and is really outside the scope of the work I was tasked to do - the testsuite was already broken if a previous run left a test database behind.

ramcq replied:
@wjt I've used our new robot overlords to generate a patch for this idea, I've tested and reviewed manually but would appreciate a 2nd opinion! https://github.com/ojwb/xapian-glib/pull/1

ojwb committed:
Update required Xapian version to 2.0.0

ojwb committed:
Wrap new QueryParser flag

ojwb commented:
Xapian 2.0.0 is now released, so I've updated the required version to that (rather than specifying a known-good commit sha).  I also wrapped a new QueryParser flag that was added between this PR and the release (so all such flags are now wrapped).
